src/axom/quest/MeshTester.hpp

```diff
@@ -33,12 +61,196 @@ enum class WatertightStatus : signed char
+  const double* z = surface_mesh->getCoordinateArray( mint::Z_COORDINATE );
+
+  for ( int n=0 ; n < 3 ; ++n )
+  {
+    const axom::IndexType nodeIdx = triCell[ n ];
+    tri[ n ][ 0 ] = x[ nodeIdx ];
+    tri[ n ][ 1 ] = y[ nodeIdx ];
+    tri[ n ][ 2 ] = z[ nodeIdx ];
+  }
+
+  return tri;
+}
 
 /// \name Mesh test and repair
 /// @{
 
 /*!
- * \brief Find self-intersections and degenerate triangles in a surface mesh.
+ * \brief Find self-intersections and degenerate triangles in a surface mesh
+ *  utilizing a Bounding Volume Hierarchy.
+ *
+ * \param [in] surface_mesh A triangle mesh in three dimensions
+ * \param [out] intersection Pairs of indices of intersecting mesh triangles
+ * \param [out] degenerateIndices indices of degenerate mesh triangles
+ *
+ * After running this function over a surface mesh, intersection will be filled
+ * with pairs of indices of intersecting triangles and degenerateIndices will
+ * be filled with the indices of the degenerate triangles in the mesh.
+ * Triangles that share vertex pairs (adjacent triangles in a watertight
+ * surface mesh) are not reported as intersecting.  Degenerate triangles
+ * are not reported as intersecting other triangles.
+ *
+ */
+#if defined(AXOM_USE_RAJA)
+template < typename ExecSpace, typename FloatType >
+void findTriMeshIntersectionsBVH(
+  mint::UnstructuredMesh< mint::SINGLE_SHAPE >* surface_mesh,
+  std::vector<std::pair<int, int> > & intersections,
+  std::vector<int> & degenerateIndices)
+{
+  SLIC_INFO("Running BVH intersection algorithm "
+            << " in execution Space: "
+            << axom::execution_space< ExecSpace >::name());
+
+  // Get allocator
+  int allocatorID = axom::execution_space< ExecSpace >::allocatorID();
+  axom::setDefaultAllocator( allocatorID );
+
+  constexpr int NDIMS = 3;
+  constexpr int stride = 2 * NDIMS;
+  const int ncells = surface_mesh->getNumberOfCells();
+
+  Triangle3* tris = axom::allocate <Triangle3> (ncells);
+
+  FloatType* xmin = axom::allocate< FloatType >( ncells );
+  FloatType* ymin = axom::allocate< FloatType >( ncells );
+  FloatType* zmin = axom::allocate< FloatType >( ncells );
+
+  FloatType* xmax = axom::allocate< FloatType >( ncells );
+  FloatType* ymax = axom::allocate< FloatType >( ncells );
+  FloatType* zmax = axom::allocate< FloatType >( ncells );
+
+  // Each access-aligned bounding box represented by 2 (x,y,z) points
+  FloatType* aabbs = axom::allocate< FloatType >( ncells * stride );
+
+  // Finding degenerate indices and bounding boxes from Triangles
+  Triangle3 t1 = Triangle3();
+  Triangle3 t2 = Triangle3();
+
+  for (int i=0 ; i < ncells ; i++)
+  {
+    t1=getMeshTriangle(i, surface_mesh);
+    tris[i] = t1;
+
+    if (t1.degenerate())
+    {
+      degenerateIndices.push_back(i);
+    }
+
+    SpatialBoundingBox triBB = compute_bounds(t1);
+
+    const IndexType offset = i * stride;
+
+    xmin[i] = aabbs[ offset ]     = triBB.getMin()[0];
```

Comment:
```cpp
     xmin[i] = aabbs[ offset ]     = triBB.getMin()[0]; 
```

Are these arrays on the host as well?

Reply:
I don't see explicit memory copies.  Is this function using unified memory? That might be another source of inefficiency for the GPU version.  

If so, I don't think it would take too much effort with axom's memory manager (using umpire) to use explicit memory transfers on a few arrays.

If not, I am not sure how the results are getting transferred.  Perhaps the bounding box bug is somehow related to this?

Reply:
This code is definitely relying on Unified memory. 

Unfortunately, using unified memory will still have memory motion even if the data is not accessed on the host, it depends on the driver/runtime, so we cannot really rely on it. 

You may be able to create a unified memory allocator with hints, which may offer some benefits.

My suggestion is to:

1. Add annotations (we now have annotation macros) and look a the profile with `nvvp`

2. Use a device memory pool and use explicit memory copies back to the host (this would have tremendous impact in performance)

3. Pass the pool allocator when constructing the BVH so that the BVH will use the pool allocator internally as well.

Reply:
Thanks @gzagaris -- found it -- https://github.com/LLNL/axom/blob/0ed1af3682b6fdf71bc08c4788f94aa0d54993f1/src/axom/core/execution/internal/cuda_exec.hpp#L60

Reply:
> My suggestion is to:
> 
> 
> Add annotations (we now have annotation macros) and look a the profile with nvvp
> 
> 
> Use a device memory pool and use explicit memory copies back to the host (this would have tremendous impact in performance)
> 
> 
> Pass the pool allocator when constructing the BVH so that the BVH will use the pool allocator internally as well.

Thanks for the suggestion!
Here was the profile of the `findTriMeshIntersectionsBVH` function using Unified memory:

![nvvp mesh tester no pool](https://user-images.githubusercontent.com/53839862/90432948-ed01ad80-e07f-11ea-9cbe-54089309a4a5.PNG)

I added a device pool to reduce the memory motion:
 
![nvvp mesh tester device pool](https://user-images.githubusercontent.com/53839862/90433133-3225df80-e080-11ea-8f71-59ab8a1ded28.PNG)

Reply:
Outstanding!